server/data/data-library/src/main/java/org/apache/james/domainlist/lib/AbstractDomainList.java

```diff
@@ -158,21 +159,23 @@ public boolean containsDomain(String domain) throws DomainListException {
         // https://stackoverflow.com/questions/37919648/concatenating-immutablelists
         // A work-around is to use Iterables.concat() until something like
         // https://github.com/google/guava/issues/1029 is implemented.
-        ImmutableList<String> detectedIps = detectIps(Iterables.concat(domains, detectedDomains));
-        Iterable<String> allDomains = Iterables.concat(domains, detectedDomains, detectedIps);
+        Iterable<String> domainsWithoutIp = (detectedDomains.isEmpty() ? domains : Iterables.concat(domains, detectedDomains));
+        ImmutableList<String> detectedIps = detectIps(domainsWithoutIp);
+        ImmutableList<String> allDomains = ImmutableList.copyOf(detectedIps.isEmpty() ? domainsWithoutIp : Iterables.concat(domainsWithoutIp, detectedIps));
```

Comment:
I would be in favour of always doing the concatenation:
 - Having a condition is misleading
 - Concant with an empty collection is not expensive.

Reply:
Done